src/wallet/mod.rs

```diff
@@ -619,9 +619,25 @@ where
             _ => 1,
         };
 
+        let last_sync_height = self
+            .database()
+            .get_sync_time()?
+            .map(|sync_time| sync_time.block_time.height);
+
+        // If they didn't tell us the current height, we assume it's the latest sync height.
+        let current_height = params.current_height.or(last_sync_height);
+
         let lock_time = match params.locktime {
-            // No nLockTime, default to 0
-            None => requirements.timelock.unwrap_or(0),
+            // When no nLockTime is specified, we try to prevent fee sniping, if possible
+            None => {
+                // Fee sniping can be partially prevented by setting the timelock
+                // to current_height. If we don't know the current_height,
+                // we default to 0.
+                let fee_sniping_height = current_height.unwrap_or(0);
```

Comment:
Tested ACK. Pulled the code locally, tested on branch, merged on master and tested. @danielabrozzoni I know this PR passed the design phase but I was wondering why on fee sniping prevention you are setting the timelock to the `current_height ` and not `current_height + 1` as in Bitcoin Core ([check last paragraph of this section](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch07.asciidoc#timelock-defense-against-fee-sniping)).

Reply:
That's an excellent question! I have to admit that when I started working on this, I was pretty sure Core would set it to `current_height + 1`. After implementing it tho, I noticed that if I set it to `current_height + 1`, then nodes wouldn't relay my transaction (as the nLockTime wasn't expired yet)!
After digging a bit into Bitcoin Core, I found out that they, as well, set the nLockTime to the best height:

https://github.com/bitcoin/bitcoin/blob/45d8b1e94a8e3e68f210d496330f5e16219d2a8d/src/wallet/spend.cpp#L633

But then, they improve privacy by setting the nLockTime even further back in the past sometimes:
https://github.com/bitcoin/bitcoin/blob/45d8b1e94a8e3e68f210d496330f5e16219d2a8d/src/wallet/spend.cpp#L639-L641

Seems like Mastering Bitcoin is wrong... There's already an issue open https://github.com/bitcoinbook/bitcoinbook/issues/997, I hope they'll eventually fix :)

Reply:
Thank you for sharing your learnings with me!